differs/rpm_diff.go

```diff
@@ -364,3 +366,115 @@ func unlock() error {
 	daemonMutex.Unlock()
 	return nil
 }
+
+type RPMLayerAnalyzer struct {
+}
+
+// Name returns the name of the analyzer.
+func (a RPMLayerAnalyzer) Name() string {
+	return "RPMLayerAnalyzer"
+}
+
+// Diff compares the installed rpm packages of image1 and image2 for each layer
+func (a RPMLayerAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := singleVersionLayerDiff(image1, image2, a)
+	return diff, err
+}
+
+// Analyze collects information of the installed rpm packages on each layer
+func (a RPMLayerAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := singleVersionLayerAnalysis(image, a)
+	return analysis, err
+}
+
+// getPackages returns an array of maps of installed rpm packages on each layer
+func (a RPMLayerAnalyzer) getPackages(image pkgutil.Image) ([]map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	var packages []map[string]util.PackageInfo
+	if _, err := os.Stat(path); err != nil {
+		// invalid image directory path
+		return packages, err
+	}
+
+	// try to find the rpm binary in bin/ or usr/bin/
+	rpmBinary := filepath.Join(path, "bin/rpm")
+	if _, err := os.Stat(rpmBinary); err != nil {
+		rpmBinary = filepath.Join(path, "usr/bin/rpm")
+		if _, err = os.Stat(rpmBinary); err != nil {
+			logrus.Errorf("Could not detect RPM binary in unpacked image %s", image.Source)
+			return packages, nil
+		}
+	}
+
+	packages, err := rpmDataFromLayerFS(image)
+	if err != nil {
+		logrus.Info("Running RPM binary from image in a container")
+		return rpmDataFromLayeredContainers(image.Image)
+	}
+	return packages, err
+}
+
+// rpmDataFromLayerFS runs a local rpm binary, if any, to query the layer
+// rpmdb and returns an array of maps of installed packages.
+func rpmDataFromLayerFS(image pkgutil.Image) ([]map[string]util.PackageInfo, error) {
```

Comment:
The logic in this function is really similar to `rpmDataFromImageFS`. Could you pull out some of the shared code between them and reuse it in both functions? Apart from iterating over layers here I think they're almost identical.

Reply:
I was also thinking about it, I can try to move some part to another function an see how it looks like. I think only lines 435->444 are clear candidates to be moved to another function.